pkg/devicetemplateconverter/profile_fetcher.go

```diff
@@ -0,0 +1,72 @@
+// Copyright © 2022 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicetemplateconverter
+
+import (
+	"context"
+
+	"go.thethings.network/lorawan-stack/v3/pkg/cluster"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"google.golang.org/grpc"
+)
+
+// Component abstracts the underlying *component.Component.
+type Component interface {
+	GetPeerConn(ctx context.Context, role ttnpb.ClusterRole, ids cluster.EntityIdentifiers) (*grpc.ClientConn, error)
+	AllowInsecureForCredentials() bool
+}
+
+// ProfileFetcher handles the validation and fetching of end device profile in the Device Repository.
+type ProfileFetcher struct {
+	Component        Component
+	DeviceRepository ttnpb.DeviceRepositoryClient
+}
+
+// ShouldFetchProfile dictactes if the end device has the necessary fields to fetch its profile.
+func (*ProfileFetcher) ShouldFetchProfile(device *ttnpb.EndDevice) bool {
+	return device.GetVersionIds().GetVendorId() != 0
+}
+
+// FetchProfile provides the end device profile.
+func (pf *ProfileFetcher) FetchProfile(ctx context.Context, device *ttnpb.EndDevice) (*ttnpb.EndDeviceTemplate, error) {
+	if pf.DeviceRepository != nil {
+		return pf.DeviceRepository.GetTemplate(ctx, &ttnpb.GetTemplateRequest{
+			EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
+				VendorId:        device.GetVersionIds().GetVendorId(),
+				VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
+			},
+		})
+	}
+
+	conn, err := pf.Component.GetPeerConn(ctx, ttnpb.ClusterRole_DEVICE_REPOSITORY, nil)
+	if err != nil {
+		log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
+		return nil, err
+	}
+
+	opt, err := rpcmetadata.WithForwardedAuth(ctx, pf.Component.AllowInsecureForCredentials())
+	if err != nil {
+		return nil, err
+	}
+
+	return ttnpb.NewDeviceRepositoryClient(conn).GetTemplate(ctx, &ttnpb.GetTemplateRequest{
+		EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
+			VendorId:        device.GetVersionIds().GetVendorId(),
+			VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
+		},
+	}, opt)
```

Comment:
Duplication of `GetTemplate` call isn't necessary. 
You're basically switching on the client connection so only that needs to be conditional.

Reply:
Not so sure about this, if I were to do something like:
```go
client, err := func() (ttnpb.DeviceRepositoryClient, error) {
if pf.DeviceRepository != nil {
	return pf.DeviceRepository, nil
}

conn, err := pf.Component.GetPeerConn(ctx, ttnpb.ClusterRole_DEVICE_REPOSITORY, nil)
if err != nil {
	log.FromContext(ctx).WithError(err).Warn("Failed to get Device Repository peer")
	return nil, err
}

return ttnpb.NewDeviceRepositoryClient(conn), nil
}()
if err != nil {
return nil, err
}

opt, err := rpcmetadata.WithForwardedAuth(ctx, pf.Component.AllowInsecureForCredentials())
if err != nil {
return nil, err
}
return client.GetTemplate(ctx, &ttnpb.GetTemplateRequest{
EndDeviceProfileIds: &ttnpb.EndDeviceProfileIdentifiers{
	VendorId:        device.GetVersionIds().GetVendorId(),
	VendorProfileId: device.GetVersionIds().GetVendorProfileId(),
},
}, opt)
```
There is a situation in which we are injecting the `WithForwardedAuth` into my request, which isn't a big problem but a bit I'm not sure if I should be forcing that into the test.